_posts/2022-01-01-the-kubernetes-discovery-cache-blessing-and-curse.md

```diff
@@ -0,0 +1,101 @@
+
+Controllers like [crossplane.io](https://crossplane.io), [GCP Config Connector](https://github.com/GoogleCloudPlatform/k8s-config-connector) or [Azure Service Operator](https://github.com/Azure/azure-service-operator) have adopted the pattern to a large degree and install 100s, if not 1,000s of CRDs on clusters. However that doesn't come without its drawbacks...
+
+<!--more-->
+
+These draw backs aren't due to high load on the Kubernetes API server. In fact, that actually has a pretty robust and advanced rate limiting mechanism through the [priority and fairness](https://kubernetes.io/docs/concepts/cluster-administration/flow-control) design, that most likely will ensure that the API server doesn't crash even if a lot of requests are made.
+
+However, installing many CRDs on a cluster can impact the [OpenAPI spec publishing](https://github.com/crossplane/crossplane/issues/2649), as well as the [discovery cache creation](https://github.com/kubernetes/kubectl/issues/1126). OpenAPI spec creation for many CRDs has [recently been fixed](https://github.com/kubernetes/kube-openapi/pull/251) through implementing lazy marshalling. While an interesting concept, this could be the topic of another blog post as in this one we are focusing on the latter one: discovery cache creation.
+
+It started with more and more log message like the following when using regular `kubectl` commands:
+
+```text
+Waited for 1.140352693s due to client-side throttling, not priority and fairness, request: GET:https://10.212.0.242/apis/storage.k8s.io/v1beta1?timeout=32s
+```
+
+What's interesting is that that immediately excludes the priority and fairness mechanism described earlier and talks about 'client-side throttling'. My first instinct however was just to suppress the log line because I hadn't asked `kubectl` to print any debug logs for instance with `-v 1`. I found [this issue on kubernetes/kubernetes](https://github.com/kubernetes/kubernetes/pull/101634) pursuing the same goal and gave it a thumbs up in the hope to just suppress this annoying log message that you couldn't switch off. However as the discussion on that PR progressed and specifically [this comment](https://github.com/kubernetes/kubernetes/pull/101634#issuecomment-840691885), saying that "this log message saves 10 hours of debugging for every hour it costs someone trying to hide it", got me thinking that there must be more to the story and that just not printing the log message is not going to be the right approach. The PR eventually was closed without merging.
```

Comment:
my first instinct, however, was...

Reply:
for instance, with (add commas)

Reply:
check the sentence pursuing the same goal

Reply:
However, comma

Reply:
as the pr progressed, and specificallt this comment? confused, did you see this comment ? check this sentence

Reply:
there must be more to the story and that merely not printing the log message is not the right approach. don't like my substitution either but check this sentence it is confusing

Reply:
Once it's rendered it will look like this:

However, as the discussion on that PR progressed and specifically [this comment](https://github.com/kubernetes/kubernetes/pull/101634#issuecomment-840691885), saying that "this log message saves 10 hours of debugging for every hour it costs someone trying to hide it", got me thinking that there must be more to the story and that merely not printing the log message is not the right approach.